mqtt/mqtt-bridge/src/pump.rs

```diff
@@ -0,0 +1,170 @@
+#![allow(dead_code)]
+use std::{cell::RefCell, rc::Rc};
+
+use futures_util::{
+    future::{select, Either, FutureExt},
+    pin_mut, select,
+    stream::{StreamExt, TryStreamExt},
+};
+use tokio::sync::{oneshot, oneshot::Receiver};
+use tracing::debug;
+use tracing::error;
+
+use mqtt3::PublishHandle;
+
+use crate::{
+    bridge::BridgeError,
+    client::{ClientShutdownHandle, MqttClient},
+    message_handler::MessageHandler,
+    persist::{MessageLoader, PublicationStore, WakingMemoryStore},
+};
+
+#[derive(Debug, Clone)]
+pub enum PumpType {
+    Local,
+    Remote,
+}
+
+pub struct Pump {
+    client: MqttClient<MessageHandler<WakingMemoryStore>>,
+    client_shutdown: ClientShutdownHandle,
+    publish_handle: PublishHandle,
+    subscriptions: Vec<String>,
+    loader: Rc<RefCell<MessageLoader<WakingMemoryStore>>>,
+    persist: PublicationStore<WakingMemoryStore>,
+    pump_type: PumpType,
+    bridge_name: String,
+}
+
+impl Pump {
+    pub fn new(
+        client: MqttClient<MessageHandler<WakingMemoryStore>>,
+        subscriptions: Vec<String>,
+        loader: Rc<RefCell<MessageLoader<WakingMemoryStore>>>,
+        persist: PublicationStore<WakingMemoryStore>,
+        pump_type: PumpType,
+        bridge_name: String,
+    ) -> Result<Self, BridgeError> {
+        let publish_handle = client
+            .publish_handle()
+            .map_err(BridgeError::PublishHandle)?;
+        let client_shutdown = client.shutdown_handle()?;
+
+        Ok(Self {
+            client,
+            client_shutdown,
+            publish_handle,
+            subscriptions,
+            loader,
+            persist,
+            pump_type,
+            bridge_name,
+        })
+    }
+
+    pub async fn subscribe(&mut self) -> Result<(), BridgeError> {
+        self.client
+            .subscribe(&self.subscriptions)
+            .await
+            .map_err(BridgeError::Subscribe)?;
+
+        Ok(())
+    }
+
+    pub async fn run(&mut self, shutdown: Receiver<()>) {
+        debug!("starting pumps for {} bridge...", self.bridge_name);
+
+        let (loader_shutdown, loader_shutdown_rx) = oneshot::channel::<()>();
+        let publish_handle = self.publish_handle.clone();
+        let persist = self.persist.clone();
+        let loader = self.loader.clone();
+        let mut client_shutdown = self.client_shutdown.clone();
+        let bridge_name = self.bridge_name.clone();
+        let pump_type = self.pump_type.clone();
+
+        // egress pump
+        let f1 = async move {
+            // TODO REVIEW: Is panicking fine?
+            let mut loader_borrow = loader.borrow_mut();
```

Comment:
Is potential panic ok here? If panic then the thread will stop and the edgehub container will restart.

Reply:
Actually, we should probably just exit from this async block so the pump shuts down and logs an error. That way we can handle however we want.

Reply:
Denis suggested that we can use borrow and then our tests can catch any bad panic cases if they arise.